src/fold/traverse.py

```diff
@@ -0,0 +1,42 @@
+from copy import deepcopy
+from typing import Callable, List
+
+from .base import Composite, Optimizer, Pipeline, Transformation
+from .utils.list import empty_if_none
+
+
+def traverse_apply(pipeline: Pipeline, apply_func: Callable) -> Pipeline:
+    def _traverse_apply(pipeline):
+        if isinstance(pipeline, List):
+            return [_traverse_apply(t) for t in pipeline]
+        elif isinstance(pipeline, Composite):
+            composite = pipeline
+            return composite.clone(clone_child_transformations=_traverse_apply)
+        elif isinstance(pipeline, Optimizer):
+            raise ValueError("Optimizer is not supported within an Optimizer.")
+        elif isinstance(pipeline, Transformation):
+            return apply_func(pipeline)
+
+    return _traverse_apply(deepcopy(pipeline))
```

Comment:
I moved the `deepcopying` here rather than two lines above at the apply_func, so that the whole pipeline is first copied.